File: examples/02_form_fill_and_submit.mjs
#!/usr/bin/env node
/**
 * Example 2 — Auto-fill a form, click submit, wait for network, export results.
 *
 * Usage:
 *     browsercli start --dir /tmp/demo-site
 *     cd clients/node && npm run build && cd ../..
 *     node examples/02_form_fill_and_submit.mjs
 *     browsercli stop
 */

import { writeFileSync } from "node:fs";
import { join } from "node:path";
import { setTimeout } from "node:timers/promises";

import {
  BrowserCLI,
  SessionError,
  ConnectionError,
  ServerError,
} from "../clients/node/dist/src/index.js";

async function main() {
  let ac;
  try {
    ac = BrowserCLI.connect();
  } catch (err) {
    if (err instanceof SessionError) {
      console.log(`ERROR: ${err.message}`);
      console.log(
        "Hint: start the daemon first with: browsercli start --dir /tmp/demo-site"
      );
      return;
    }
    throw err;
  }

  try {
    const info = await ac.status();
    const serveDir = info.dir;
    if (!serveDir) {
      console.log("ERROR: daemon has no serve directory");
      return;
    }

    // 1. Write a page with a form that "submits" via JS (no real server needed)
    const html = `<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8"><title>Form Demo</title></head>
<body>
  <h1>Registration</h1>
  <form id="reg-form" onsubmit="return handleSubmit(event)">
    <input id="name" type="text" placeholder="Name" />
    <input id="email" type="email" placeholder="Email" />
    <button id="submit-btn" type="submit">Submit</button>
  </form>
  <div id="result" style="display:none"></div>
  <script>
    function handleSubmit(e) {
      e.preventDefault();
      var name = document.getElementById('name').value;
      var email = document.getElementById('email').value;
      // Simulate an API call with fetch to a local echo
      fetch('/echo?name=' + encodeURIComponent(name) + '&email=' + encodeURIComponent(email))
        .catch(function() {})  // will 404, that's fine
        .finally(function() {
          var el = document.getElementById('result');
          el.textContent = 'Submitted: ' + name + ' <' + email + '>';
          el.style.display = 'block';
          console.log('form-submitted', {name: name, email: email});
        });
      return false;
    }
  </script>
</body>
</html>
`;
    writeFileSync(join(serveDir, "index.html"), html);
    await setTimeout(1000);
    await ac.goto("/");
    await setTimeout(500);

    // 2. Fill the form
    await ac.domType("#name", "Alice Agent", true);
    await ac.domType("#email", "alice@example.com", true);
    console.log("Filled form fields");

    // 3. Clear network log, then click submit
    await ac.network(0, true);
    await ac.domClick("#submit-btn");
    await setTimeout(1000);

    // 4. Check the result text appeared
    const resultText = await ac.domQuery("#result", "text");
    console.log(`Result element: ${resultText}`);

    // 5. Inspect network log for the fetch call
    const entries = await ac.network();
    console.log(`Network entries after submit: ${entries.length}`);
    for (const e of entries) {
      console.log(`  ${e.method} ${e.url} -> ${e.status}`);
    }

    // 6. Export results as JSON
    const report = {
      form_result: resultText,
      network_requests: entries,
    };
    const outPath = join(serveDir, "form_report.json");
    writeFileSync(outPath, JSON.stringify(report, null, 2));
    console.log(`Report saved to ${outPath}`);
  } catch (err) {
    if (err instanceof ConnectionError || err instanceof ServerError) {
      console.log(`ERROR: ${err.message}`);
    } else {
      throw err;
    }
  }
}

main();

File: examples/02_form_fill_and_submit.py
#!/usr/bin/env python3
"""Example 2 — Auto-fill a form, click submit, wait for network, export results.

Usage:
    browsercli start --dir /tmp/demo-site
    python examples/02_form_fill_and_submit.py
    browsercli stop
"""

import json
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "clients" / "python"))

from browsercli import BrowserCLI, SessionError, ConnectionError, ServerError


def main() -> None:
    try:
        ac = BrowserCLI.connect()
    except SessionError as e:
        print(f"ERROR: {e}")
        print("Hint: start the daemon first with: browsercli start --dir /tmp/demo-site")
        return

    try:
        info = ac.status()
        serve_dir = info.get("dir", "")
        if not serve_dir:
            print("ERROR: daemon has no serve directory")
            return

        # 1. Write a page with a form that "submits" via JS (no real server needed)
        html = """\
<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8"><title>Form Demo</title></head>
<body>
  <h1>Registration</h1>
  <form id="reg-form" onsubmit="return handleSubmit(event)">
    <input id="name" type="text" placeholder="Name" />
    <input id="email" type="email" placeholder="Email" />
    <button id="submit-btn" type="submit">Submit</button>
  </form>
  <div id="result" style="display:none"></div>
  <script>
    function handleSubmit(e) {
      e.preventDefault();
      var name = document.getElementById('name').value;
      var email = document.getElementById('email').value;
      // Simulate an API call with fetch to a local echo
      fetch('/echo?name=' + encodeURIComponent(name) + '&email=' + encodeURIComponent(email))
        .catch(function() {})  // will 404, that's fine
        .finally(function() {
          var el = document.getElementById('result');
          el.textContent = 'Submitted: ' + name + ' <' + email + '>';
          el.style.display = 'block';
          console.log('form-submitted', {name: name, email: email});
        });
      return false;
    }
  </script>
</body>
</html>
"""
        Path(serve_dir, "index.html").write_text(html)
        time.sleep(1)
        ac.goto("/")
        time.sleep(0.5)

        # 2. Fill the form
        ac.dom_type("#name", "Alice Agent", clear=True)
        ac.dom_type("#email", "alice@example.com", clear=True)
        print("Filled form fields")

        # 3. Clear network log, then click submit
        ac.network(clear=True)
        ac.dom_click("#submit-btn")
        time.sleep(1)

        # 4. Check the result text appeared
        result_text = ac.dom_query("#result", mode="text")
        print(f"Result element: {result_text}")

        # 5. Inspect network log for the fetch call
        entries = ac.network()
        print(f"Network entries after submit: {len(entries)}")
        for entry in entries:
            print(f"  {entry['method']} {entry['url']} -> {entry['status']}")

        # 6. Export results as JSON
        report = {
            "form_result": result_text,
            "network_requests": entries,
        }
        out_path = str(Path(serve_dir) / "form_report.json")
        Path(out_path).write_text(json.dumps(report, indent=2))
        print(f"Report saved to {out_path}")

    except (ConnectionError, ServerError) as e:
        print(f"ERROR: {e}")


if __name__ == "__main__":
    main()
